File: ProjC1/src/projc1/Game/GameCanvas.java
package projc1.Game;
import java.awt.*;
import java.awt.event.*;
import javax.swing.*;

public class GameCanvas extends JPanel implements KeyListener, ActionListener {
    // AGREGACIÓN: GameCanvas tiene estos objetos pero pueden existir independientemente
    private ScoreSystem scoreSystem;
    private RenderEngine renderEngine;
    private PositionGenerator positionGenerator;
    
    // COMPOSICIÓN: Estos objetos son parte integral del juego
    private Player player;
    private Coin coin;
    private Timer timer;
    
    public GameCanvas(String imageRoute) {
        initializeGame(imageRoute);
        setupUI();
    }
    
    private void initializeGame(String imageRoute) {
        // AGREGACIÓN: Creamos objetos que podrían existir independientemente
        scoreSystem = new ScoreSystem();
        renderEngine = new RenderEngine();
        positionGenerator = new PositionGenerator();
        
        // COMPOSICIÓN: Estos objetos son parte integral del juego
        player = new Player(50, 50);
        player.setPlayerImage(imageRoute);

        coin = null;
        
        generateNewCoin();
        
        // Timer para el game loop
        timer = new Timer(16, this);
        timer.start();
    }
    
    private void setupUI() {
        setBackground(Color.BLACK);
        setFocusable(true);
        addKeyListener(this);
        
        addMouseListener(new MouseAdapter() {
            @Override
            public void mouseClicked(MouseEvent e) {
                requestFocusInWindow();
            }
        });
    }
    
    // DEPENDENCIA: Usa PositionGenerator para crear nueva moneda
    private void generateNewCoin() {
        int maxX = getWidth() - 20;
        int maxY = getHeight() - 20;
        
        if (maxX <= 0) maxX = 350;
        if (maxY <= 0) maxY = 250;
        
        // DEPENDENCIA: Usamos PositionGenerator
        Position coinPosition = positionGenerator.generateRandomPosition(
            maxX, maxY, player.getPosition(), 50);
        
        coin = new Coin(coinPosition.getX(), coinPosition.getY());
    }
    
    @Override
    protected void paintComponent(Graphics g) {
        super.paintComponent(g);
        
        Graphics2D g2d = (Graphics2D) g;
        g2d.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);
        
        // ASOCIACIÓN: Usar RenderEngine para dibujar
        if (coin != null) {
            coin.draw(g2d, renderEngine);
        }
        
        player.draw(g2d, renderEngine);
        
        // Dibujar UI usando RenderEngine
        Font titleFont = new Font("Arial", Font.BOLD, 14);
        Font instructionFont = new Font("Arial", Font.PLAIN, 12);
        
        renderEngine.drawText(g2d, "Puntuación: " + scoreSystem.getScore(), 
                             10, 25, titleFont, Color.WHITE);
        renderEngine.drawText(g2d, "Monedas: " + scoreSystem.getCoinsCollected(), 
                             10, 45, titleFont, Color.WHITE);
        

    }
    
    private void checkCollisions() {
        if (coin != null && coin.intersects(player.getPosition(), player.getSize())) {
            // AGREGACIÓN: Usar ScoreSystem
            scoreSystem.addScore(coin.getValue());
            generateNewCoin();
        }
    }
    
    @Override
    public void keyPressed(KeyEvent e) {
        int key = e.getKeyCode();
        
        switch (key) {
            case KeyEvent.VK_LEFT:
            case KeyEvent.VK_A:
                player.moveLeft(getWidth());
                break;
            case KeyEvent.VK_RIGHT:
            case KeyEvent.VK_D:
                player.moveRight(getWidth());
                break;
            case KeyEvent.VK_UP:
            case KeyEvent.VK_W:
                player.moveUp();
                break;
            case KeyEvent.VK_DOWN:
            case KeyEvent.VK_S:
                player.moveDown(getHeight());
                break;

        }
        repaint();
    }
    
    @Override
    public void keyReleased(KeyEvent e) {}
    
    @Override
    public void keyTyped(KeyEvent e) {}
    
    @Override
    public void actionPerformed(ActionEvent e) {
        checkCollisions();
        repaint();
    }
}

File: ProcC1_pyver/game/game_canvas.py
import tkinter as tk
from tkinter import font as tkfont


from game.components.position_generator import PositionGenerator
from game.components.render_engine import RenderEngine
from game.components.score_system import ScoreSystem
from game.entities.coin import Coin
from game.entities.player import Player

class GameCanvas(tk.Canvas):
    def __init__(self, parent, image_path):
        super().__init__(parent, bg="black", width=400, height=300)
        
        # Agregación
        self.score_system = ScoreSystem()
        self.render_engine = RenderEngine()
        self.position_generator = PositionGenerator()
        
        # Composición
        self.player = Player(50, 50)
        self.player.set_player_image(image_path)
        
        self.coin = None
        self.generate_new_coin()
        
        # Configuración de eventos
        self.bind("<KeyPress>", self.on_key_press)
        self.focus_set()
        
        # Game loop
        self.after(16, self.update_game)
    
    def generate_new_coin(self):
        max_x = self.winfo_width() - 20 if self.winfo_width() > 20 else 380
        max_y = self.winfo_height() - 20 if self.winfo_height() > 20 else 280
        
        coin_pos = self.position_generator.generate_random_position(
            max_x, max_y, self.player.get_position(), 50)
        
        if coin_pos:
            self.coin = Coin(coin_pos.get_x(), coin_pos.get_y())
    
    def draw(self):
        self.delete("all")
        
        # Dibujar moneda
        if self.coin:
            self.coin.draw(self, self.render_engine)
        
        # Dibujar jugador
        self.player.draw(self, self.render_engine)
        
        # Dibujar UI
        title_font = tkfont.Font(family="Arial", size=12, weight="bold")
        self.render_engine.draw_text(
            self, f"Puntuación: {self.score_system.get_score()}", 
            10, 20, title_font, "white")
        self.render_engine.draw_text(
            self, f"Monedas: {self.score_system.get_coins_collected()}", 
            10, 40, title_font, "white")
    
    def check_collisions(self):
        if (self.coin and 
            self.coin.intersects(self.player.get_position(), self.player.get_size())):
            self.score_system.add_score(self.coin.get_value())
            self.generate_new_coin()
    
    def on_key_press(self, event):
        key = event.keysym
        
        if key in ("Left", "a"):
            self.player.move_left(self.winfo_width())
        elif key in ("Right", "d"):
            self.player.move_right(self.winfo_width())
        elif key in ("Up", "w"):
            self.player.move_up()
        elif key in ("Down", "s"):
            self.player.move_down(self.winfo_height())
        
        self.draw()
    
    def update_game(self):
        self.check_collisions()
        self.draw()
        self.after(16, self.update_game)